File: OsmAnd-java/src/main/java/net/osmand/router/PrecalculatedRouteDirection.java
package net.osmand.router;

import java.util.ArrayList;
import java.util.List;

import gnu.trove.list.array.TIntArrayList;
import net.osmand.binary.RouteDataObject;
import net.osmand.data.LatLon;
import net.osmand.data.QuadPointDouble;
import net.osmand.data.QuadRect;
import net.osmand.data.QuadTree;
import net.osmand.util.MapUtils;

public class PrecalculatedRouteDirection {
	
	private int[] pointsX;
	private int[] pointsY;
	private float minSpeed;
	private float maxSpeed;
	private float[] tms;
	private boolean followNext;
	private static final int SHIFT = (1 << (31 - 17));
	private static final int[] SHIFTS = new int[]{1 << (31 - 15), 1 << (31 - 13), 1 << (31 - 12), 
		1 << (31 - 11), 1 << (31 - 7)};
	
	private List<Integer> cachedS = new ArrayList<Integer>();
	
	private long startPoint = 0;
	private long endPoint = 0;
//	private DataTileManager<Integer> indexedPoints = new DataTileManager<Integer>(17);
	QuadTree<Integer> quadTree = new QuadTree<Integer>(new QuadRect(0, 0, Integer.MAX_VALUE, Integer.MAX_VALUE),
			8, 0.55f);
	private float startFinishTime;
	private float endFinishTime; 
	
	public PrecalculatedRouteDirection(TIntArrayList px, TIntArrayList py, List<Float> speedSegments, float maxSpeed) {
		this.maxSpeed = maxSpeed;
		init(px, py, speedSegments);
	}
	
	private PrecalculatedRouteDirection(List<RouteSegmentResult> ls, float maxSpeed) {
		this.maxSpeed = maxSpeed;
		init(ls);
	}
	
	private PrecalculatedRouteDirection(LatLon[] ls, float maxSpeed) {
		this.maxSpeed = maxSpeed;
		init(ls);
	}
	
	private PrecalculatedRouteDirection(PrecalculatedRouteDirection parent, int s1, int s2) {
		this.minSpeed = parent.minSpeed;
		this.maxSpeed = parent.maxSpeed;
		boolean inverse = false;
		if (s1 > s2) {
			int tmp = s1;
			s1 = s2;
			s2 = tmp;
			inverse = true;
		}
		tms = new float[s2 - s1 + 1];
		pointsX = new int[s2 - s1 + 1];
		pointsY = new int[s2 - s1 + 1];
		for (int i = s1; i <= s2; i++) {
			int shiftInd = i - s1;
			pointsX[shiftInd] = parent.pointsX[i];
			pointsY[shiftInd] = parent.pointsY[i];
//			indexedPoints.registerObjectXY(parent.pointsX.get(i), parent.pointsY.get(i), pointsX.size() - 1);
			quadTree.insert(shiftInd, parent.pointsX[i], parent.pointsY[i]);
			tms[shiftInd] = parent.tms[i] - parent.tms[inverse ? s1 : s2];
		}
	}
	
	public static PrecalculatedRouteDirection build(List<RouteSegmentResult> ls, float cutoffDistance, float maxSpeed){
		int begi = 0;
		float d = cutoffDistance;
		for (; begi < ls.size(); begi++) {
			d -= ls.get(begi).getDistance();
			if (d < 0) {
				break;
			}
		}
		int endi = ls.size();
		d = cutoffDistance;
		for (; endi > 0; endi--) {
			d -= ls.get(endi - 1).getDistance();
			if (d < 0) {
				break;
			}
		}
		if(begi < endi) {
			return new PrecalculatedRouteDirection(ls.subList(begi, endi), maxSpeed);
		}
		return null;
	}
	
	public static PrecalculatedRouteDirection build(LatLon[] ls, float maxSpeed){
		return new PrecalculatedRouteDirection(ls, maxSpeed);
	}
	

	private void init(List<RouteSegmentResult> ls) {
		TIntArrayList px = new TIntArrayList();
		TIntArrayList py = new TIntArrayList();
		List<Float> speedSegments = new ArrayList<Float>();
		for (RouteSegmentResult s : ls) {
			boolean plus = s.getStartPointIndex() < s.getEndPointIndex();
			int i = s.getStartPointIndex();
			RouteDataObject obj = s.getObject();
			float routeSpd = (s.getRoutingTime() == 0 || s.getDistance() == 0) ? maxSpeed : 
				(s.getDistance() / s.getRoutingTime());
			while (true) {
				i = plus? i + 1 : i -1;
				px.add(obj.getPoint31XTile(i));
				py.add(obj.getPoint31YTile(i));
				speedSegments.add(routeSpd);
				if (i == s.getEndPointIndex()) {
					break;
				}
			}
		}
		init(px, py, speedSegments);
	}
	
	private void init(LatLon[] ls) {
		TIntArrayList px = new TIntArrayList();
		TIntArrayList py = new TIntArrayList();
		List<Float> speedSegments = new ArrayList<Float>();
		for (LatLon s : ls) {
			float routeSpd = maxSpeed; // (s.getDistance() / s.getRoutingTime())
			px.add(MapUtils.get31TileNumberX(s.getLongitude()));
			py.add(MapUtils.get31TileNumberY(s.getLatitude()));
			speedSegments.add(routeSpd);
		}
		init(px, py, speedSegments);
	}

	private void init(TIntArrayList px, TIntArrayList py, List<Float> speedSegments) {
		float totaltm = 0;
		List<Float> times = new ArrayList<Float>();
		for (int i = 0; i < px.size(); i++) {
			// MapUtils.measuredDist31 vs BinaryRoutePlanner.squareRootDist
			// use measuredDist31 because we use precise s.getDistance() to calculate routeSpd
			int ip = i == 0 ? 0 : i - 1;
			float dist = (float) MapUtils.measuredDist31(px.get(ip), py.get(ip), px.get(i), py.get(i));
			float tm = dist / speedSegments.get(i);// routeSpd;
			times.add(tm);
			quadTree.insert(i, px.get(i), py.get(i));
			// indexedPoints.registerObjectXY();
			totaltm += tm;
		}
		pointsX = px.toArray();
		pointsY = py.toArray();
		tms = new float[times.size()];
		float totDec = totaltm;
		for(int i = 0; i < times.size(); i++) {
			totDec -= times.get(i);
			tms[i] = totDec;
		}
	}

	public float timeEstimate(int sx31, int sy31, int ex31, int ey31) {
		long l1 = calc(sx31, sy31);
		long l2 = calc(ex31, ey31);
		int x31;
		int y31;
		boolean start;
		if (l1 == startPoint || l1 == endPoint) {
			start = l1 == startPoint;
			x31 = ex31;
			y31 = ey31;
		} else if(l2 == startPoint || l2 == endPoint) {
			start = l2 == startPoint;
			x31 = sx31;
			y31 = sy31;
		} else {
			throw new UnsupportedOperationException();
		}
		int ind = getIndex(x31, y31);
		if(ind == -1) {
			return -1;
		}
		if((ind == 0 && start) || 
				(ind == pointsX.length - 1 && !start)) {
			return -1;
		}
		float distToPoint = getDeviationDistance(x31, y31, ind);
		float deviationPenalty = distToPoint / minSpeed;
		float finishTime = (start? startFinishTime : endFinishTime);
		if(start) {
			return (tms[0] - tms[ind]) +  deviationPenalty + finishTime;
		} else {
			return tms[ind] + deviationPenalty + finishTime;
		}
	}
	
	public float getDeviationDistance(int x31, int y31) {
		int ind = getIndex(x31, y31);
		if(ind == -1) {
			return 0;
		}
		return getDeviationDistance(x31, y31, ind);
	}

	public float getDeviationDistance(int x31, int y31, int ind) {
		float distToPoint = 0; //BinaryRoutePlanner.squareRootDist(x31, y31, pointsX.get(ind), pointsY.get(ind));
		if(ind < pointsX.length - 1 && ind != 0) {
			double nx = BinaryRoutePlanner.squareRootDist(x31, y31, pointsX[ind + 1], pointsY[ind + 1]);
			double pr = BinaryRoutePlanner.squareRootDist(x31, y31, pointsX[ind - 1], pointsY[ind - 1]);
			int nind =  nx > pr ? ind -1 : ind +1;
			QuadPointDouble proj = MapUtils.getProjectionPoint31(x31, y31, pointsX[ind], pointsY[ind], pointsX[nind], pointsX[nind]);
			distToPoint = (float) BinaryRoutePlanner.squareRootDist(x31, y31, (int)proj.x, (int)proj.y) ;
		}
		return distToPoint;
	}

	public int getIndex(int x31, int y31) {
		int ind = -1;
		cachedS.clear();
//		indexedPoints.getObjects(x31 - SHIFT, y31 - SHIFT, x31 + SHIFT, y31 + SHIFT, cachedS);
		quadTree.queryInBox(new QuadRect(x31 - SHIFT, y31 - SHIFT, x31 + SHIFT, y31 + SHIFT), cachedS);
		if (cachedS.size() == 0) {
			for (int k = 0; k < SHIFTS.length; k++) {
				quadTree.queryInBox(new QuadRect(x31 - SHIFTS[k], y31 - SHIFTS[k], x31 + SHIFTS[k], y31 + SHIFTS[k]), cachedS);
//				indexedPoints.getObjects(x31 - SHIFTS[k], y31 - SHIFTS[k], x31 + SHIFTS[k], y31 + SHIFTS[k],cachedS);
				if (cachedS.size() != 0) {
					break;
				}
			}
			if (cachedS.size() == 0) {
				return -1;
			}
		}
		double minDist = 0;
		for (int i = 0; i < cachedS.size(); i++) {
			Integer n = cachedS.get(i);
			double ds = BinaryRoutePlanner.squareRootDist(x31, y31, pointsX[n], pointsY[n]);
			if (ds < minDist || i == 0) {
				ind = n;
				minDist = ds;
			}
		}
		return ind;
	}

	private long calc(int x31, int y31) {
		return ((long) x31) << 32l + ((long)y31);
	}

	public void setFollowNext(boolean followNext) {
		this.followNext = followNext;
	}

	public boolean isFollowNext() {
		return followNext;
	}

	public PrecalculatedRouteDirection adopt(RoutingContext ctx) {
		int ind1 = getIndex(ctx.startX, ctx.startY);
		int ind2 = getIndex(ctx.targetX, ctx.targetY);
		minSpeed = ctx.getRouter().getDefaultSpeed();
		maxSpeed = ctx.getRouter().getMaxSpeed();
		if(ind1 == -1) {
			return null;
		}
		if(ind2 == -1) {
			return null;
		}
		PrecalculatedRouteDirection routeDirection = new PrecalculatedRouteDirection(this, ind1, ind2);
		routeDirection.startPoint = calc(ctx.startX, ctx.startY);
		routeDirection.startFinishTime = (float) BinaryRoutePlanner.squareRootDist(pointsX[ind1], pointsY[ind1], ctx.startX, ctx.startY) / maxSpeed;
		routeDirection.endPoint = calc(ctx.targetX, ctx.targetY);
		routeDirection.endFinishTime = (float) BinaryRoutePlanner.squareRootDist(pointsX[ind2], pointsY[ind2], ctx.targetX, ctx.targetY) / maxSpeed;
		routeDirection.followNext = followNext;
		return routeDirection;
	}

	public void updatePreciseStartEnd(int sx, int sy, int ex, int ey) {
		if (sx > 0 && sy > 0) {
			int ind = getIndex(sx, sy);
			if (ind != -1) {
				startPoint = calc(sx, sy);
				startFinishTime = (float) BinaryRoutePlanner.squareRootDist(pointsX[ind], pointsY[ind], sx, sy) / maxSpeed;
			}
		}
		if (ex > 0 && ey > 0) {
			int ind = getIndex(ex, ey);
			if (ind != -1) {
				endPoint = calc(ex, ey);
				endFinishTime = (float) BinaryRoutePlanner.squareRootDist(pointsX[ind], pointsY[ind], ex, ey) / maxSpeed;
			}
		}
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/routing/PrecalculatedRouteDirection.kt
package net.osmand.shared.routing

import net.osmand.shared.data.KLatLon
import net.osmand.shared.data.KQuadRect
import net.osmand.shared.data.KQuadTree
import net.osmand.shared.util.KMapUtils
import net.osmand.shared.util.collections.KTIntArrayList
import kotlin.jvm.JvmStatic

/**
 * A route that has already been found, kept so the next search can lean on it.
 *
 * The points are indexed by place and carry the time still left to the end, so the A* heuristic can
 * ask "how long from here, if I follow the route I already have" instead of guessing from straight
 * line distance. That is what makes recalculating a long route cheap: the search stays in a corridor
 * around the old one, and only the ends are worked out again.
 *
 * The C++ core reads the eight fields below - the two coordinate arrays, the times, the speeds and
 * the finish times - and runs the same estimate itself.
 */
class PrecalculatedRouteDirection {

	private lateinit var pointsX: IntArray
	private lateinit var pointsY: IntArray
	private var minSpeed: Float = 0f
	private var maxSpeed: Float = 0f
	private lateinit var tms: FloatArray
	private var followNext: Boolean = false

	private val cachedS: MutableList<Int> = ArrayList()

	private var startPoint: Long = 0
	private var endPoint: Long = 0
	private val quadTree = KQuadTree<Int>(
		KQuadRect(0.0, 0.0, Int.MAX_VALUE.toDouble(), Int.MAX_VALUE.toDouble()), 8, 0.55f
	)
	private var startFinishTime: Float = 0f
	private var endFinishTime: Float = 0f

	constructor(px: KTIntArrayList, py: KTIntArrayList, speedSegments: List<Float>, maxSpeed: Float) {
		this.maxSpeed = maxSpeed
		initPoints(px, py, speedSegments)
	}

	private constructor(ls: List<RouteSegmentResult>, maxSpeed: Float) {
		this.maxSpeed = maxSpeed
		initFromRoute(ls)
	}

	private constructor(ls: Array<KLatLon>, maxSpeed: Float) {
		this.maxSpeed = maxSpeed
		initFromLatLon(ls)
	}

	private constructor(parent: PrecalculatedRouteDirection, start: Int, end: Int) {
		this.minSpeed = parent.minSpeed
		this.maxSpeed = parent.maxSpeed
		var s1 = start
		var s2 = end
		var inverse = false
		if (s1 > s2) {
			val tmp = s1
			s1 = s2
			s2 = tmp
			inverse = true
		}
		tms = FloatArray(s2 - s1 + 1)
		pointsX = IntArray(s2 - s1 + 1)
		pointsY = IntArray(s2 - s1 + 1)
		for (i in s1..s2) {
			val shiftInd = i - s1
			pointsX[shiftInd] = parent.pointsX[i]
			pointsY[shiftInd] = parent.pointsY[i]
			quadTree.insert(shiftInd, parent.pointsX[i].toFloat(), parent.pointsY[i].toFloat())
			tms[shiftInd] = parent.tms[i] - parent.tms[if (inverse) s1 else s2]
		}
	}

	private fun initFromRoute(ls: List<RouteSegmentResult>) {
		val px = KTIntArrayList()
		val py = KTIntArrayList()
		val speedSegments = ArrayList<Float>()
		for (s in ls) {
			val plus = s.getStartPointIndex() < s.getEndPointIndex()
			var i = s.getStartPointIndex()
			val obj = s.getObject()
			val routeSpd = if (s.getRoutingTime() == 0f || s.getDistance() == 0f) maxSpeed
			else (s.getDistance() / s.getRoutingTime())
			while (true) {
				i = if (plus) i + 1 else i - 1
				px.add(obj.getPoint31XTile(i))
				py.add(obj.getPoint31YTile(i))
				speedSegments.add(routeSpd)
				if (i == s.getEndPointIndex()) {
					break
				}
			}
		}
		initPoints(px, py, speedSegments)
	}

	private fun initFromLatLon(ls: Array<KLatLon>) {
		val px = KTIntArrayList()
		val py = KTIntArrayList()
		val speedSegments = ArrayList<Float>()
		for (s in ls) {
			val routeSpd = maxSpeed // (s.getDistance() / s.getRoutingTime())
			px.add(KMapUtils.get31TileNumberX(s.longitude))
			py.add(KMapUtils.get31TileNumberY(s.latitude))
			speedSegments.add(routeSpd)
		}
		initPoints(px, py, speedSegments)
	}

	private fun initPoints(px: KTIntArrayList, py: KTIntArrayList, speedSegments: List<Float>) {
		var totaltm = 0f
		val times = ArrayList<Float>()
		for (i in 0 until px.size()) {
			// KMapUtils.measuredDist31 vs BinaryRoutePlanner.squareRootDist
			// use measuredDist31 because we use precise s.getDistance() to calculate routeSpd
			val ip = if (i == 0) 0 else i - 1
			val dist = KMapUtils.measuredDist31(px[ip], py[ip], px[i], py[i]).toFloat()
			val tm = dist / speedSegments[i] // routeSpd
			times.add(tm)
			quadTree.insert(i, px[i].toFloat(), py[i].toFloat())
			totaltm += tm
		}
		pointsX = px.toArray()
		pointsY = py.toArray()
		tms = FloatArray(times.size)
		var totDec = totaltm
		for (i in times.indices) {
			totDec -= times[i]
			tms[i] = totDec
		}
	}

	fun timeEstimate(sx31: Int, sy31: Int, ex31: Int, ey31: Int): Float {
		val l1 = calc(sx31, sy31)
		val l2 = calc(ex31, ey31)
		val x31: Int
		val y31: Int
		val start: Boolean
		if (l1 == startPoint || l1 == endPoint) {
			start = l1 == startPoint
			x31 = ex31
			y31 = ey31
		} else if (l2 == startPoint || l2 == endPoint) {
			start = l2 == startPoint
			x31 = sx31
			y31 = sy31
		} else {
			val sInd = getIndex(sx31, sy31)
			val eInd = getIndex(ex31, ey31)
			val err = "startPoint:$startPoint, endPoint:$endPoint, l1:$l1, l2:$l2, sx31:$sx31," +
					" sy31:$sy31, ex31:$ex31, ey31:$ey31, startIndex: $sInd, endIndex:$eInd"
			throw UnsupportedOperationException(err)
		}
		val ind = getIndex(x31, y31)
		if (ind == -1) {
			return -1f
		}
		if ((ind == 0 && start) || (ind == pointsX.size - 1 && !start)) {
			return -1f
		}
		val distToPoint = getDeviationDistance(x31, y31, ind)
		val deviationPenalty = distToPoint / minSpeed
		val finishTime = if (start) startFinishTime else endFinishTime
		return if (start) {
			(tms[0] - tms[ind]) + deviationPenalty + finishTime
		} else {
			tms[ind] + deviationPenalty + finishTime
		}
	}

	fun getDeviationDistance(x31: Int, y31: Int): Float {
		val ind = getIndex(x31, y31)
		if (ind == -1) {
			return 0f
		}
		return getDeviationDistance(x31, y31, ind)
	}

	fun getDeviationDistance(x31: Int, y31: Int, ind: Int): Float {
		var distToPoint = 0f
		if (ind < pointsX.size - 1 && ind != 0) {
			val nx = KMapUtils.squareRootDist31(x31, y31, pointsX[ind + 1], pointsY[ind + 1])
			val pr = KMapUtils.squareRootDist31(x31, y31, pointsX[ind - 1], pointsY[ind - 1])
			val nind = if (nx > pr) ind - 1 else ind + 1
			// pointsX twice is how java had it; the projection is only a penalty, and changing it
			// would move every route that leans on a precalculated one
			val proj = KMapUtils.getProjectionPoint31(
				x31, y31, pointsX[ind], pointsY[ind], pointsX[nind], pointsX[nind]
			)
			distToPoint = KMapUtils.squareRootDist31(x31, y31, proj.x.toInt(), proj.y.toInt()).toFloat()
		}
		return distToPoint
	}

	fun getIndex(x31: Int, y31: Int): Int {
		var ind = -1
		cachedS.clear()
		quadTree.queryInBox(
			KQuadRect(
				(x31 - SHIFT).toDouble(), (y31 - SHIFT).toDouble(),
				(x31 + SHIFT).toDouble(), (y31 + SHIFT).toDouble()
			), cachedS
		)
		if (cachedS.size == 0) {
			for (k in SHIFTS.indices) {
				quadTree.queryInBox(
					KQuadRect(
						(x31 - SHIFTS[k]).toDouble(), (y31 - SHIFTS[k]).toDouble(),
						(x31 + SHIFTS[k]).toDouble(), (y31 + SHIFTS[k]).toDouble()
					), cachedS
				)
				if (cachedS.size != 0) {
					break
				}
			}
			if (cachedS.size == 0) {
				return -1
			}
		}
		var minDist = 0.0
		for (i in cachedS.indices) {
			val n = cachedS[i]
			val ds = KMapUtils.squareRootDist31(x31, y31, pointsX[n], pointsY[n])
			if (ds < minDist || i == 0) {
				ind = n
				minDist = ds
			}
		}
		return ind
	}

	private fun calc(x31: Int, y31: Int): Long {
		return (x31.toLong() shl 32) + y31.toLong()
	}

	fun setFollowNext(followNext: Boolean) {
		this.followNext = followNext
	}

	fun isFollowNext(): Boolean = followNext

	/**
	 * Cuts the part of this route between the two points out, and measures how far the real start
	 * and target sit from it.
	 *
	 * Takes the points and the router rather than a routing context: the context holds an open obf
	 * reader and stays in OsmAnd-java, and this is all of it that was ever read here.
	 */
	fun adopt(startX: Int, startY: Int, targetX: Int, targetY: Int, router: VehicleRouter): PrecalculatedRouteDirection? {
		val ind1 = getIndex(startX, startY)
		val ind2 = getIndex(targetX, targetY)
		minSpeed = router.getDefaultSpeed()
		maxSpeed = router.getMaxSpeed()
		if (ind1 == -1) {
			return null
		}
		if (ind2 == -1) {
			return null
		}
		val routeDirection = PrecalculatedRouteDirection(this, ind1, ind2)
		routeDirection.startPoint = calc(startX, startY)
		routeDirection.startFinishTime =
			(KMapUtils.squareRootDist31(pointsX[ind1], pointsY[ind1], startX, startY) / maxSpeed).toFloat()
		routeDirection.endPoint = calc(targetX, targetY)
		routeDirection.endFinishTime =
			(KMapUtils.squareRootDist31(pointsX[ind2], pointsY[ind2], targetX, targetY) / maxSpeed).toFloat()
		routeDirection.followNext = followNext
		return routeDirection
	}

	fun updatePreciseStartEnd(sx: Int, sy: Int, ex: Int, ey: Int) {
		if (sx > 0 && sy > 0) {
			val ind = getIndex(sx, sy)
			if (ind != -1) {
				startPoint = calc(sx, sy)
				startFinishTime = (KMapUtils.squareRootDist31(pointsX[ind], pointsY[ind], sx, sy) / maxSpeed).toFloat()
			}
		}
		if (ex > 0 && ey > 0) {
			val ind = getIndex(ex, ey)
			if (ind != -1) {
				endPoint = calc(ex, ey)
				endFinishTime = (KMapUtils.squareRootDist31(pointsX[ind], pointsY[ind], ex, ey) / maxSpeed).toFloat()
			}
		}
	}

	companion object {

		private const val SHIFT = 1 shl (31 - 17)
		private val SHIFTS = intArrayOf(
			1 shl (31 - 15), 1 shl (31 - 13), 1 shl (31 - 12),
			1 shl (31 - 11), 1 shl (31 - 7)
		)

		@JvmStatic
		fun build(ls: List<RouteSegmentResult>, cutoffDistance: Float, maxSpeed: Float): PrecalculatedRouteDirection? {
			var begi = 0
			var d = cutoffDistance
			while (begi < ls.size) {
				d -= ls[begi].getDistance()
				if (d < 0) {
					break
				}
				begi++
			}
			var endi = ls.size
			d = cutoffDistance
			while (endi > 0) {
				d -= ls[endi - 1].getDistance()
				if (d < 0) {
					break
				}
				endi--
			}
			if (begi < endi) {
				return PrecalculatedRouteDirection(ls.subList(begi, endi), maxSpeed)
			}
			return null
		}

		@JvmStatic
		fun build(ls: Array<KLatLon>, maxSpeed: Float): PrecalculatedRouteDirection =
			PrecalculatedRouteDirection(ls, maxSpeed)
	}
}
